Add support for configuration chaining

HaoK opened the pull request:
cc @divega

HaoK commented:
Fixes https://github.com/aspnet/Configuration/issues/339

davidfowl commented on src/Microsoft.Extensions.Configuration/ConfigurationBuilder.cs:
```diff
@@ -58,6 +59,31 @@ public IConfigurationBuilder Add(IConfigurationProvider provider, bool load)
+        /// </summary>
+        /// <param name="configurationBuilder">The <see cref="IConfigurationBuilder"/> to add to.</param>
+        /// <param name="config">The <see cref="IConfiguration"/> to include.</param>
+        /// <param name="key">The key of the root of the configuartion to inculde.</param>
```
tpyo inculde :smile:

davidfowl commented on src/Microsoft.Extensions.Configuration/ChainedConfigurationProvider.cs:
```diff
@@ -0,0 +1,38 @@
+{
+    public class ChainedConfigurationProvider : ConfigurationProvider
+    {
+        private readonly IConfiguration _config;
```
I want to make you type more so _configuration and configuration

HaoK replied:
gross...good thing there's ctrl space

davidfowl commented on src/Microsoft.Extensions.Configuration/ConfigurationBuilder.cs:
```diff
@@ -58,6 +59,31 @@ public IConfigurationBuilder Add(IConfigurationProvider provider, bool load)
+        /// <param name="config">The <see cref="IConfiguration"/> to include.</param>
+        /// <param name="key">The key of the root of the configuartion to inculde.</param>
+        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
+        public IConfigurationBuilder AddConfiguration(IConfiguration config, string key)
```
Hmmm do we really need this if you can call GetSection on your own?

HaoK replied:
It seems like low value sugar, but it was requested for in the initial issue so I just added it.

davidfowl replied:
Ok

davidfowl commented on test/Microsoft.Extensions.Configuration.Test/ConfigurationTest.cs:
```diff
@@ -12,6 +12,86 @@ namespace Microsoft.Extensions.Configuration.Test
+            var configurationBuilder = new ConfigurationBuilder();
+
+            // Act
+            configurationBuilder.Add(memConfigSrc1, load: false);
```
What does `load:false` do?

HaoK replied:
@divega probably needs to confirm, I just copied this from another unit test, I would guess if load is false, the data is lazy loaded and only loaded on first call to get as opposed to loaded up front?  It looks like Load is only called by Reload in ConfigurationRoot, but that can't be right for this to be working

divega replied:
I didn't actually know this, but `Load()` is a no-op in the base `ConfigurationProvider` and `MemoryConfigurationProvider` doesn't bother overriding it, so the `load: false` used in the test is completely superfluous.

At first glance it doesn't look like anything but unit tests is using this overload of `ConfigurationBuilder.Add()` and BTW `IConfigurationBuilder` doesn't define it. 

 I filed #351 to remove it.

DamianEdwards commented on src/Microsoft.Extensions.Configuration/ChainedConfigurationProvider.cs:
```diff
@@ -0,0 +1,38 @@
+    {
+        private readonly IConfiguration _config;
+
+        public ChainedConfigurationProvider(IConfiguration config)
```
Where's the prefix support? https://github.com/aspnet/benchmarks/blob/master/src/Benchmarks/ConfigurationConfigurationProvider.cs#L49

DamianEdwards replied:
Note the original issue states there should be an overload that accepts a prefix to use as a filter.

davidfowl replied:
In the extension method.

HaoK commented:
Updated: more urations, and moved methods into extension methods so they show up on the interface instead of only on ConfigurationBuilder

divega commented on src/Microsoft.Extensions.Configuration/ConfigurationBuilderExtensions.cs:
```diff
@@ -7,9 +7,42 @@
+        /// <param name="config">The <see cref="IConfiguration"/> to include.</param>
+        /// <param name="key">The key of the root of the configuartion to include.</param>
+        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
+        public static IConfigurationBuilder AddConfiguration(this IConfigurationBuilder configurationBuilder, IConfiguration configuration, string key)
```
I think the signature should actually be as @DamianEdwards requested in #339:

``` C#
public static IConfigurationBuilder AddConfiguration(this IConfigurationBuilder builder, string prefix, IConfiguration configuration);
```

divega replied:
Do not confuse the `prefix` argument with a `key` to get a subsection. From what I remember the prefix was supposed to be pre-pended to the keys of the underlying IConfiguration to produce the keys of the wrapping configuration, e.g.:

``` C#
hostingConfiguration["X"] = "Y";
var appConfiguration = new ConfigurationBuilder().AddConfiguration("hosting_",hostingConfiguration).Build();
Assert.Equal(appConfiguration["hosting_X"], "X"); 
``` 

Unfortunately I don't see any notes on this in the request and I can't find any notes in the related PR https://github.com/aspnet/Hosting/pull/448, but hopefully @DamianEdwards, @davidfowl or @Tratcher can clarify.

HaoK replied:
@DamianEdwards can you provide some more details about how you expect the prefix overload to behave?

DamianEdwards replied:
I linked to working code: https://github.com/aspnet/benchmarks/blob/dev/src/Benchmarks/ConfigurationConfigurationProvider.cs

HaoK replied:
Cool, so few questions about the prototype.  It looks like you only are importing top level values, i.e. app.Key = value, and I don't think this would pickup child sections like app.Foo:Bar:Key = value.  I think we should support this if we are doing this as a general feature.

Related to this, if we moved away from app. Prefix, and instead made it a  specific section, i.e. app:EnableDbTests = true, then this would map to importing exactly one config section (perhaps eliminating the need for the prefix logic/overload... i.e. builder.Include(env.Configuration.GetSection("app"))

DamianEdwards replied:
Yeah I think that works (assume importing of a section rather than relying
on a prefix).

On Tue, Dec 15, 2015 at 3:29 PM, Hao Kung <notifications@github.com> wrote:

> In
> src/Microsoft.Extensions.Configuration/ConfigurationBuilderExtensions.cs
> <https://github.com/aspnet/Configuration/pull/349#discussion_r47712539>:
>
> > +            if (configurationBuilder == null)
> > +            {
> > +                throw new ArgumentNullException(nameof(configurationBuilder));
> > +            }
> > +            configurationBuilder.Add(new ChainedConfigurationProvider(configuration));
> > +            return configurationBuilder;
> > +        }
> > +
> > +        /// <summary>
> > +        /// Includes an existing IConfiguration as a configuration provider to <paramref name="configuraton"/>.
> > +        /// </summary>
> > +        /// <param name="configurationBuilder">The <see cref="IConfigurationBuilder"/> to add to.</param>
> > +        /// <param name="config">The <see cref="IConfiguration"/> to include.</param>
> > +        /// <param name="key">The key of the root of the configuartion to include.</param>
> > +        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
> > +        public static IConfigurationBuilder AddConfiguration(this IConfigurationBuilder configurationBuilder, IConfiguration configuration, string key)
>
> Cool, so few questions about the prototype. It looks like you only are
> importing top level values, i.e. app.Key = value, and I don't think this
> would pickup child sections like app.Foo:Bar:Key = value. I think we should
> support this if we are doing this as a general feature.
>
> Related to this, if we moved away from app. Prefix, and instead made it a
> specific section, i.e. app:EnableDbTests = true, then this would map to
> importing exactly one config section (perhaps eliminating the need for the
> prefix logic/overload... i.e.
> builder.Include(env.Configuration.GetSection("app"))
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/aspnet/Configuration/pull/349/files#r47712539>.
>



-- 
Cheers,
Damian

divega commented on src/Microsoft.Extensions.Configuration/ConfigurationBuilderExtensions.cs:
```diff
@@ -7,9 +7,42 @@
+        /// <param name="configurationBuilder">The <see cref="IConfigurationBuilder"/> to add to.</param>
+        /// <param name="config">The <see cref="IConfiguration"/> to include.</param>
+        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
+        public static IConfigurationBuilder AddConfiguration(this IConfigurationBuilder configurationBuilder, IConfiguration configuration)
```
Having `configurationBuilder.AddConfiguration()` is a bit puzzling IMO. I would suggest using a name that conveys the chainedness, e.g. `AddChainedConfiguration()`.

davidfowl replied:
Ugh

divega replied:
What does `Ugh` mean in this context?

davidfowl replied:
Don't like the name. AddConfigurationSource?

divega replied:
Yep, I am sure we can come up with something better :smile:

HaoK replied:
Since we are going away from the other extensions naming pattern.  How about something like Copy/Include/Import depending on exactly what kind of intent we want to imply.

divega replied:
That said, `Include()` and `Import()` seem decent candidates to me. So does `AddExistingConfiguration()`.

Note that we are not necessarily moving away from then pattern we use for the other providers, e.g. the name `AddChainedConfiguration()` I proposed happened to align with the name of the provider `ChainedConfigurationProvider` in this PR. If we don't like that maybe we still can fine a pair of nicer names for the method and the class that still align.

divega commented on src/Microsoft.Extensions.Configuration/ChainedConfigurationProvider.cs:
```diff
@@ -0,0 +1,38 @@
+    {
+        private readonly IConfiguration _configuration;
+
+        public ChainedConfigurationProvider(IConfiguration configuration)
```
If I am right, you will need a constructor that accepts a `prefix` string.

divega replied:
(And off course you will need to revisit all the rest of the code in this class to concatenate the prefix :smile:)

davidfowl commented on src/Microsoft.Extensions.Configuration/ChainedConfigurationProvider.cs:
```diff
@@ -0,0 +1,38 @@
+            _configuration = configuration;
+        }
+
+        public override void Load()
```
Is the only way to handle this via copying? There's no other API that is lazier?

divega replied:
I think we can be lazier if it is important. It would most likely mean we have to implement `IConfigurationProvider` on our own rather than leveraging the base implementation in `ConfigurationProvider`.

divega replied:
BTW, what should happen if the chained configuration is reloaded? Is it ok for the IConfigurations to return a mix of fresh and stale data from different providers?

Should have a way to propagate the reload like we do for file sources?

davidfowl replied:
Yes

davidfowl replied:
I'm not sure about a mix of stale and fresh data. How does that all work today? It feels like this is the wrong way to implement this tbh. Copying all of the data over kinda sucks and wastes memory.

divega replied:
Today all the providers I have looked at that represent an external source (i.e. not the in-memory one) eagerly copy all the data from the source into their own Data dictionary on `Load()`. So if the source changes Data becomes stale until you ask the configuration root to reload. `FileProvoderExtensions.ReloadOnChanged()` then can be use to automate the reloads.

Tratcher commented on src/Microsoft.Extensions.Configuration/ConfigurationBuilderExtensions.cs:
```diff
@@ -7,9 +7,42 @@
 namespace Microsoft.Extensions.Configuration
 {
-    public static class MemoryConfigurationExtensions
+    public static class ConfigurationBuilderExtensions
```
Don't name extension method classes after the thing they extend, you end up with weird conflicts. Name them after the functionality you're adding.

divega replied:
Good point. Hint: memory and chained are built-in providers.

HaoK replied:
Since there were two different providers, I didn't have a great name, DefaultConfigurationBuilderExtensions is the latest stab... feel free to propose something less lame :)

divega commented on src/Microsoft.Extensions.Configuration/ChainedConfigurationProvider.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace Microsoft.Extensions.Configuration
+{
+    public class ChainedConfigurationProvider : ConfigurationProvider
```
Once we decide the best name for the Add extension method, try to make the name of the class align (or the other way around).

HaoK commented:
Updated with some feedback, new extension method is `configBuilder.Include(config)`.

Renamed ChainedConfigurationProvider => IncludeConfigurationProvider.

Can we use config sections(`app:enableDebug=true`) instead of name prefixes (`app.enableDebug=true`)?  That eliminates the need for any prefix filtering business.  I kept the overload which takes a sectionKey for now.

divega commented on src/Microsoft.Extensions.Configuration/IncludeConfigurationProvider.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace Microsoft.Extensions.Configuration
+{
+    public class IncludeConfigurationProvider : ConfigurationProvider
```
Is the naming brainstorming still open? I have a slight preference for `Import()` and `ImportedConfigurationProvider`. Otherwise I think this class should probably be called `IncludedConfigurationProvider` (notice the "d" that is currently missing).

Caveat: English is my second language

HaoK replied:
Yeah these are easy to change, for me I probably rank them.  

1. Include
2. Copy
3. Import

I think its clearer for the name to exactly match the extension method that adds it, since these classes aren't used directly

divega replied:
> I think its clearer for the name to exactly match the extension method that adds it, since these classes aren't used directly

Users still get exposed to this class classes even if less frequently than the extension method. We should pick a name that makes some sense.

divega commented on src/Microsoft.Extensions.Configuration/IncludeConfigurationProvider.cs:
```diff
@@ -0,0 +1,31 @@
+            }
+        }
+
+        private void AddSection(IConfigurationSection section, string keyPrefix)
```
Nit: this is not a `keyPrefix` but a `parentKey`. It helps given that we give "prefix" a different meaning elsewhere.

divega commented on src/Microsoft.Extensions.Configuration/IncludeConfigurationProvider.cs:
```diff
@@ -0,0 +1,31 @@
+            Data.Add(keyPrefix, section.Value);
+            foreach (var child in section.GetChildren())
+            {
+                AddSection(child, keyPrefix + Constants.KeyDelimiter + child.Key);
```
It is a bit of a shame that we have to rebuild the key like this every time with `keyPrefix + Constants.KeyDelimiter + child.Key`. I.e. child.Key does some processing every time and then we do the concatenation. I think we can avoid most of it (although I haven't tested it):

- If the IConfiguration passed to the constructor was an IConfigurationRoot, then `keyPrefix + Constants.KeyDelimiter + child.Key` can be replaced with just `child.Path`. 
- Otherwise it should be just `child.Path.Substring(source.Path.Lenght + 1)`, and `source.Path.Lenght + 1` remains the same for the whole provider instance so we can compute it only once.

HaoK replied:
Yeah good idea, that seems to work.  The cast for the root smells a bit.  I think this is another example where it'd be nicer to have Path/Key in the IConfiguration and simply handle the Root to have empty/null Path/Key.  Doing is casts/ checks in traversal code is pretty gross.  I updated the PR moving Path/key to the base and having them return "".  I am willing to bet you are going to have some pushback but its worth a shot :)

divega commented on src/Microsoft.Extensions.Configuration.Abstractions/IConfiguration.cs:
```diff
@@ -12,6 +12,17 @@ namespace Microsoft.Extensions.Configuration
         /// <summary>
+        /// Gets the key this element occupies in its parent or "" for the IConfigurationRoot.
+        /// </summary>
+        string Key { get; }
```
@haok moving `Key` and `Path` to `IConfiguration` is a redesign of the interfaces and I don't think it belongs in this PR. If you want to start a discussion about redesigning the interfaces let's do that separately. 

For some context, `Key`, `Path` and `Value` are defined only for "child" `IConfiguration` (i.e. `IConfigurationSection`) because they felt very unnatural on the configuration root. This was raised by @davidfowl and others. I have raised the fact that it is not great to have to always try cast/is check `IConfiguration` instances in library code to see what they support but so far this is still the design.

FWIW, if end up changing the design, we should probably move `Value` as well and get rid of the `IConfigurationSection` interface altogether.

HaoK replied:
yep, see https://github.com/aspnet/Configuration/pull/355 for a clean PR with the first stab at merging IConfigurationSection/IConfiguration

divega commented on src/Microsoft.Extensions.Configuration/IncludeConfigurationProvider.cs:
```diff
@@ -0,0 +1,40 @@
+                throw new ArgumentNullException(nameof(source));
+            }
+            int pathStart;
+            if (string.IsNullOrEmpty(source.Path))
```
Sorry if I am missing something, but `Path` is defined on `IConfigurationSection` and sources is only an `IConfiguration`, so how does this compile? :trollface:

HaoK replied:
This PR has the changes which merged IConfigurationSection already, but this can be replaced by if/cast without that PR

divega commented on src/Microsoft.Extensions.Configuration/IncludeConfigurationProvider.cs:
```diff
@@ -0,0 +1,40 @@
+            Data.Add(section.Path.Substring(pathStart), section.Value);
+            foreach (var child in section.GetChildren())
+            {
+                AddSection(child, pathStart);
```
Nit: store `pathStart` in a private field and stop passing it around.

HaoK commented:
Updated removing the merge change, IncludeConfigurationProvider => IncludedConfigurationProvider.

davidfowl commented:
Whats' the deal with this.......

HaoK commented:
Waiting for Diego to sign off. He's been out/sick. I will get this in this week or as soon as we sync up.

divega commented on src/Microsoft.Extensions.Configuration/DefaultConfigurationBuilderExtensions.cs:
```diff
@@ -7,9 +7,42 @@
+        /// <param name="sectionKey">Includes the configuration starting from the child section found with this key.</param>
+        /// <param name="config">The <see cref="IConfiguration"/> to include.</param>
+        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
+        public static IConfigurationBuilder Include(this IConfigurationBuilder configurationBuilder, string sectionKey, IConfiguration configuration)
```
Nit: consider `includedConfiguration`, `includedKey` for the parameter names.

divega commented on src/Microsoft.Extensions.Configuration/IncludedConfigurationProvider.cs:
```diff
@@ -0,0 +1,37 @@
+            }
+        }
+
+        private void AddSection(IConfigurationSection section, int pathStart)
```
Nit: consider keeping `pathStart` in a class field rather than passing it on each recursive call.

davidfowl replied:
FWIW I think it should be lazy...

HaoK replied:
So just Defer loading the data until the first time someone hits this provider?  That doesn't seem too unreasonable

divega replied:
The first time someone hits this provider will be the first time anyone asks for anything on the configuration to which the provider was added. I.e. the only case in which we save any work is if the configuration was built but never used. Not sure how much gain that is. I think @davidfowl was rather referring to not ever copying the data but instead redirecting all requests for keys or values to the underlying configuration.

HaoK replied:
So basically that means we would need to implement IConfigurationProvider ourselves and don't use the abstract base class

divega replied:
Yes, and I don't think it would be hard. `GetChildrenKeys()` looks a bit hard but in the end I think it should be easy to implement with `GetChildren()`. The only really weird thing about the contract is that `GetChildrenKeys()` takes a delimiter argument. I suspect we should loose it. I suspect if there is any coversion between the standard configuration delimiter and whatever the provider's underlying representation uses, that is implementation detail of the provider.

divega replied:
Clarification: I totally meant to say that it would *not* be hard. But my phone's keyboard wasn't as confident :smile:

HaoK replied:
Ok, filed https://github.com/aspnet/Configuration/issues/366 to track this improvement

divega replied:
OK, I think we can also have a bug to track the concern about lack of change notification/load propagation. Can you please file it and then :shipit: ?

HaoK replied:
Done https://github.com/aspnet/Configuration/issues/367

divega commented:
:shipit: but would be nice if you address the nits.

divega commented:
Got reminded just now that this won't propagate change notifications from the source configuration. Are we OK with that? Cc @lodejard @rynowak

davidfowl commented:
@divega because of the copying right?

HaoK commented:
I'll address the nits and get this in tomorrow, we can address copying/lazy in another PR if necessary

HaoK commented:
2909753f6c2aca8936a4a63c902eb9647e14f709
